feat: Diff source autodetection and caching

poliorcetics opened the pull request:
Following discussions in both https://github.com/helix-editor/helix/pull/9892 and https://github.com/helix-editor/helix/pull/8056, this PR adds diff source autodetection and caching.

Supported diff sources

- None
- Git (when the relevant feature is active)

Work left for another PR because it would clutter this one (it's possible not all of this will be implemented, those are ideas):

- Adding "File" Diff source
- Adding command to select diff source
- Adding config for diff source detection order
- Adding a statusline element to show the currently active diff source
- Adding support for JJ as an example of another vcs based diff source (see #12022)

I did not find any issue about diff source autodetection, if there is one I missed tell me or edit the PR desc to mark it as closing the issue if relevant :)

pascalkuthe commented:
if I see this right this does "autodetection" per document and turns diff provider into a per-document thing right?

I want this to be a config option in the future that can be set for the document with `:set buffer::diff_source git` for the buffer and `:set diff.provider git` globally. 

All this is part of the config refactor I am working on. I also want to add an `:unset` option so you could do `:unset buffer diff.provider` to reset to the global diff source (otherwise if you change the global setting the global setting doesn't change).

If no explicit value is provided autodetection would be used.

This would be particularly relevant to this PR because I don't want the `file` diff source to be enabled by default. It's a neat feature but should be opt-in. That would mean its precluded from the auto-detection.

I was only thinking of VCS with the autodection. I thought that there would usually only be one vcs per workspace (we use `.git` for root detection anyway) so we would just do the autodetection once at startup/when changing directories.

poliorcetics commented:
That's a lot of ideas in one comment :) !

> This would be particularly relevant to this PR because I don't want the file diff source to be enabled by default. It's a neat feature but should be opt-in. That would mean its precluded from the auto-detection.

Sure I can do that easily, I don't have a strong opinion and copied the previous MR (1)

> I thought that there would usually only be one vcs per workspace (we use .git for root detection anyway) so we would just do the autodetection once at startup/when changing directories.

I thought about that but for now I find it useless since the git repo is not kept open between diff calls. Ideally we would do it yes but I'm unsure it should be done in this MR just yet

There are also issues with updating this from external events and even just removing the git dir for example, but I think the proper solution will be to tell people to reload the diffs

> I want this to be a config option in the future that can be set for the document with :set buffer::diff_source git for the buffer and :set diff.provider git globally.

that can be done in another MR without issues I feel

--- 

I'll do (1) today, for the rest, especially keeping track of workspaces will need more work

pascalkuthe commented:
> I thought about that but for now I find it useless since the git repo is not kept open between diff calls. Ideally we would do it yes but I'm unsure it should be done in this MR just yet



I originally was thinking in the direction of just having a global config option that would be controlled by autodetection by default.  I was hoping to go that direction since I wanted to avoid adding `:diffsource` (since it would be covered by `:set buffer diff.source` in the future) But on second thought I like having a typable command for this from a ui standpoint. We can alias that to `:set buffer diff source` in the future (I will probably do the same for other commands which set stuff per document right now which will become proper config options in the future).

> There are also issues with updating this from external events and even just removing the git dir for example, but I think the proper solution will be to tell people to reload the diff

Yeah I think this should only update if people reload. Maybe something fancier once we have a filewatcher But that is still in the more distant future.

pascalkuthe commented on helix-vcs/src/lib.rs:
```diff
@@ -21,129 +16,186 @@ mod status;
+        // To force a miscompilation and remember to add the value to the list
+        match Self::None {
+            Self::None => (),
+            #[cfg(feature = "git")]
```
That cfg is one line too high

pascalkuthe commented on helix-vcs/src/lib.rs:
```diff
@@ -21,129 +16,186 @@ mod status;
+        }
+
+        // Keep it sorted alphabetically
+        &[
```
I think splitting this into two Methods where one returns a list of all variants and one turns a variant into a string would be more generally useful (I will need the latter to turn this into a config option for example and you can reuse it in the display implementation)

I really wish rust had a builtin way to iterate all vsriantes of a fieldless enum. In my own projects I have a macro for that but it doesn't work with cfg and it's a bit of a pain to use since you need to put the entoe enum kn the macro... Or maybe one day we will have macro_rules based derive macros

pascalkuthe commented on helix-view/src/document.rs:
```diff
@@ -1054,14 +1063,41 @@ impl Document {
+    fn reload_diff_data(&mut self, path: &Path) {
+        let diff_data = &mut self.diff_data;
+
+        diff_data.version_control_head = diff_data.source.get_current_head_name(path);
```
Switching a file to the file diff source should not be hiding the vcs statusline element. I think this should really just be a global thing instead of being determined for each buffer seperatly . You can then also use autosetectikn of the cwd there (like the git status picker). 

It's not an absolute must to do this in this PR but if you feel up to it, I think it would be nice to include here.

poliorcetics committed:
feat: Diff source caching and autodetection

poliorcetics commented:
I changed the approach completely:

- Removed the "File" diff source, I'll do it in a subsequent PR
- Removed the "diffsource" command since it's now useless
- **Added** diff provider caching, to avoid recomputing a git repo on each diff-related data gathering

poliorcetics commented:
I'll add the "File" diff source and the "diffsource" command in a PR on top when this one is merged since it's an orthogonal issue and would make reviewing this harder

poliorcetics commented:
Updated this MR and the dependent (#12022) on top of the workspace trust v2, I think I got things right regarding trust (anytime it changed, scrap the whole provider to reload it)
